### PR TITLE
Improve `sqlind-adjust-operator' and `sqlind-backward-syntactic-ws' to manage alignment for pipes.

### PR DESCRIPTION
**Pull request by PierreTechoueyres**
Correctly indent the line with pipe operator when a string around.

after the pipe.

Before the patch :
```sql
    select col1, col2
              || col3
             || 'some text'
                   || col6  as composed_column,
           col4
        || col5 as composed_column2
    from   my_table
    where  cond1 = fct1
               || 'another text'
    and    cond2 = 2;
```
After :
```sql
    select col1, col2
              || col3
              || 'some text'
              || col6  as composed_column,
           col4
        || col5 as composed_column2
    from   my_table
    where  cond1 = fct1
                || 'another text'
    and    cond2 = 2;
```

### CODE REVIEW

File: sql-indent.el [hunk 1]
```diff
@@ -136,10 +136,11 @@ whitespace, or at the beginning of the buffer."
       (unless (eq (point) (point-min))
 	(forward-char -1))
       (let ((pps (syntax-ppss (point))))
-	(if (nth 4 pps)                 ; inside a comment?
-	    (goto-char (nth 8 pps))     ; move to comment start than repeat
-	    (throw 'done (point))       ; done
-	    )))))
+	(cond
+	 ((or (nth 4 pps) (nth 3 pps))  ; inside a comment or a string ?
```
Comment (alex-hhh) [inline sql-indent.el]: This is not correct, as strings are not syntactic white-space, and this function should not skip over them.  I think this change will break indentation in other cases.

I understand the problem you were trying to fix, but I don't think this is the right solution.

File: sql-indent.el [hunk 2]
```diff
@@ -1752,7 +1755,10 @@ This is an indentation adjuster and needs to be added to the
     ;; consider them an operator and line up the line to the first word of the
     ;; line, not the operator
     (cond ((looking-at "\\W+")
-	   (let ((ofs (length (match-string 0))))
+	   (let* ((operator (match-string 0))
+		  (ofs (length operator)))
+	     (when (string-match "'$" operator)
+	       (setq ofs (1- ofs)))
```
Comment (alex-hhh) [inline sql-indent.el]: I think the `(forward-sexp -1)` line below is the one that needs to change.  Perhaps we will need to write our own "forward/backward-term" function.   I think all other places were `forward-sexp` is used will have the same problem.  I will think about how to best do this.

File: sql-indent.el [hunk 3]
```diff
@@ -1752,7 +1755,10 @@ This is an indentation adjuster and needs to be added to the
     ;; consider them an operator and line up the line to the first word of the
     ;; line, not the operator
     (cond ((looking-at "\\W+")
-	   (let ((ofs (length (match-string 0))))
+	   (let* ((operator (match-string 0))
+		  (ofs (length operator)))
+	     (when (string-match "'$" operator)
```
Comment (alex-hhh) [inline sql-indent.el]: This is not going to work if the string itself starts with a non word, like this:

```sql
select c
        || '++'
from my_table
```

It seems to me that the safest way is to change the `looking-at` regexp to list the operators explicitly.  SQLite seems to have the [following](https://sqlite.org/lang_expr.html) operators (I suspect Oracle has pretty much the same.  I think we can build a regexp with `regexp-opt` that just lists them.  Something like:

```emacs-lisp
(defconst sqlind-operator-regexp
  (concat (regexp-opt '("||" "*" "/" "%" "+" "-" "<<" ">>" "&" "|"
                        "<" "<=" ">" ">=" "=" "==" "!=" "<>")
                      t)
          "\\s-+"))
```

This way, we only match what we need and don't have to undo the extra match.

**Comment by PierreTechoueyres** [comment]
I missed a few points :-)
And there is even a much worse case:

```sql
select col1, 'a long line of text ending with a single word'
                                                   || col2
          || col3
          || 'some text' as composed_column,
       col4
    || col5 as composed_column2
from   my_table
where  cond1 = fct1
           || 'another text'
and    cond2 = 2;
```

**Comment by alex-hhh** [comment]
I have pushed some changes that I believe handle all the cases now.  Can you please check it? I think this Pull request can be closed, but I will wait for you to verify first.

**Comment by PierreTechoueyres** [comment]
Sorry, not yet.
The pathological case bellow indent incorrectly with:
```(add-hook 'sqlind-minor-mode-hook #'sqlind-setup-style-left)```

```sql
clear columns
set linesize 2500
-- set trimout on
-- set trimspool on

select sysdate from dual;
```
become:
```sql
clear columns
set linesize 2500
  -- set trimout on
  -- set trimspool on

  select sysdate from dual;
```

**Comment by alex-hhh** [comment]
Try now :-)

**Comment by PierreTechoueyres** [comment]
Sorry again. It break in another statement.
I try to format the joined file.
It seems to also break the indent-region command.


[truc.sql.txt](https://github.com/alex-hhh/emacs-sql-indent/files/1018895/truc.sql.txt)

**Comment by PierreTechoueyres** [comment]
FYI I get the following result:

```sql
clear columns
set linesize 2500
-- set trimout on
-- set trimspool on

select sysdate from dual;

select col1, 'a long line of text ending with a single word'
          || col2
          || col3
          || 'some text' as composed_column,
col4
  || col5 as composed_column2
from   my_table
where  cond1 = fct1
  || 'another text'
and    cond2 = 2;
```

**Comment by alex-hhh** [comment]
Oh well, one more try.  If it does not work, I'll try again tomorrow, it is too late here, and I'm making silly mistakes. :-)

**Comment by PierreTechoueyres** [comment]
Yeah you win !!!
It looks great:
```sql
clear columns
set linesize 2500
set trimout on trimspool on

select sysdate from dual;

select col1, 'a long line of text ending with a single word'
          || col2
          || col3
          || 'some text' as composed_column,
       col4
    || col5 as composed_column2
from   my_table
where  cond1 = fct1
            || 'another text'
and    cond2 = 2;
```
even the tricky ```set trimout on trimspool on``` was correctly indented.
Thank !!!

P.S.
As a side note : did you have time to look at #7 ?

**Comment by alex-hhh** [comment]
Well, its good that it works now.   I will add some unit tests for this.  For now, I will close this PR.

**Pull request closed without merging**